### DISCUSSION TITLE
Embedding mrdb

### DISCUSSION THREAD

**Discussion by jabamaus** [mruby/mruby#5982] [General]
Following on from https://github.com/mruby/mruby/discussions/5956 I have started looking at mrdb. The way I'd like to start experimenting is to embed it into my app and then start adding my own script commands to control it, and see how I get on. The trouble is it seems like mrdb was designed only to be a standalone tool. Is embedding it like I'm thinking vaguely feasible? I was hoping for an api much like the mruby api, eg mrdb_open(), mrdb_close() etc.

**Comment by matz** [comment]
Hmm, mrdb was originally developed as a third-party tool. So embedding it was not in the scope (until now).
Let me think about it.

**Comment by dearblue** [comment]
:black_cat: The following is a soliloquy. Also, I have not tried it, so none of the contents are accurate.

Python (CPython) seems to be able to debug using GDB.
https://github.com/python/cpython/blob/main/Tools/gdb/libpython.py

I also found a project that says Python scripts can be debugged with LLDB.
https://github.com/malor/cpython-lldb

Maybe we can use GDB or LLDB to display the mruby VM stack and set breakpoints.

I don't want to reserve the task, so please wait for someone else's results.

**Comment by dmajkic** [comment]
I believe the question was to implement simple script debug capability in app where mruby is embedded. 

In such a case - involving end-user, or whole user installed app in GDB or LLDB debugging is probably too much to handle.

mruby as a small and embedded component, exposes more complex apps. Like a game engine or some business app. It is natural to simply embed text-memo component in app, and be able to load mruby source script, set breakpoints, have a call stack and read current variable values on breakpoint. That would be great, as mruby-bin-debugger already exists. But it is implemented only as external binary.

**Comment by jabamaus** [comment]
For me, and I think a lot of other people, having a source level graphical debugger is absolutely essential to my development process. Whether that is remote or embedded is not so important. I would like mruby to provide library apis, based on mrdb that would allow me to develop some of my own debugging tools. Also in its current form I don't see how mrdb can debug code that depends on native code as that would not be linked into the executable...
